lib/opencensus/trace/span_builder.rb

```diff
@@ -314,52 +348,240 @@ def initialize span_context, sampled, skip_frames: 0
+                       max_string_length: nil
+          @max_attributes =
+            max_attributes || OpenCensus::Trace::Config.default_max_attributes
+          @max_stack_frames =
+            max_stack_frames ||
+            OpenCensus::Trace::Config.default_max_stack_frames
+          @max_annotations =
+            max_annotations || OpenCensus::Trace::Config.default_max_annotations
+          @max_message_events =
+            max_message_events ||
+            OpenCensus::Trace::Config.default_max_message_events
+          @max_links = max_links || OpenCensus::Trace::Config.default_max_links
+          @max_string_length =
+            max_string_length ||
+            OpenCensus::Trace::Config.default_max_string_length
         end
-      end
 
-      def links
-        @links.map do |link|
-          OpenCensus::Trace::Link.new \
-            link.trace_id,
-            link.span_id,
-            type: link.type,
-            attributes: link.attributes
+        ##
+        # Build a canonical attributes hash and the number of dropped items
+        # @private
+        #
+        def attribute_pieces attrs
+          result = {}
+          count = 0
+          attrs.each do |k, v|
+            break if @max_attributes != 0 && count == @max_attributes
+            result[k.to_s] =
+              case v
+              when Integer
+                if v >= MIN_INT && v <= MAX_INT
+                  v
+                else
+                  truncatable_string v.to_s
+                end
+              when true, false, TruncatableString
+                v
+              else
+                truncatable_string v.to_s
+              end
+            count += 1
+          end
+          [result, attrs.size - count]
+        end
+
+        ##
+        # Build a canonical stack trace and the number of dropped items
+        # @private
+        #
+        def stack_trace_pieces raw_trace
+          orig_size = raw_trace.size
+          if @max_stack_frames.zero? || orig_size <= @max_stack_frames
+            [raw_trace, 0]
+          else
+            [raw_trace[0, @max_stack_frames], orig_size - @max_stack_frames]
+          end
+        end
+
+        ##
+        # Build a canonical annotations list and the number of dropped items
+        # @private
+        #
+        def annotation_pieces raw_annotations
+          result = []
+          count = 0
+          raw_annotations.each do |ann|
+            break if @max_annotations != 0 && count == @max_annotations
+            attrs, dropped_attributes_count = attribute_pieces ann.attributes
+            result <<
+              OpenCensus::Trace::Annotation.new(
+                truncatable_string(ann.description),
+                attributes: attrs,
+                dropped_attributes_count: dropped_attributes_count,
+                time: ann.time
+              )
+            count += 1
+          end
+          [result, raw_annotations.size - count]
```

Comment:
For arrays, we don't need to keep a running count as arrays keep track of their size. This could be `raw_annotations.size - result.size`

Reply:
Removed the unnecessary tracking. Also for hashes since they also appear to keep track of their size.